test/e2e/framework/syncer.go

```diff
@@ -511,6 +515,256 @@ func (sf *syncerFixture) Start(t *testing.T) *StartedSyncerFixture {
+
+	virtualWorkspaceRawConfig := rawConfig.DeepCopy()
+	virtualWorkspaceRawConfig.Clusters["syncer"] = rawConfig.Clusters["base"].DeepCopy()
+	virtualWorkspaceRawConfig.Clusters["syncer"].Server = virtualWorkspaceURL
+	virtualWorkspaceRawConfig.Contexts["syncer"] = rawConfig.Contexts["base"].DeepCopy()
+	virtualWorkspaceRawConfig.Contexts["syncer"].Cluster = "syncer"
+	virtualWorkspaceRawConfig.Clusters["upsyncer"] = rawConfig.Clusters["base"].DeepCopy()
+	virtualWorkspaceRawConfig.Clusters["upsyncer"].Server = strings.Replace(virtualWorkspaceURL, "/services/syncer/", "/services/upsyncer/", 1)
+	virtualWorkspaceRawConfig.Contexts["upsyncer"] = rawConfig.Contexts["base"].DeepCopy()
+	virtualWorkspaceRawConfig.Contexts["upsyncer"].Cluster = "upsyncer"
+	syncerVWConfig, err := clientcmd.NewNonInteractiveClientConfig(*virtualWorkspaceRawConfig, "syncer", nil, nil).ClientConfig()
+	require.NoError(t, err)
+	syncerVWConfig = rest.AddUserAgent(rest.CopyConfig(syncerVWConfig), t.Name())
+	require.NoError(t, err)
+	upsyncerVWConfig, err := clientcmd.NewNonInteractiveClientConfig(*virtualWorkspaceRawConfig, "upsyncer", nil, nil).ClientConfig()
+	require.NoError(t, err)
+	upsyncerVWConfig = rest.AddUserAgent(rest.CopyConfig(upsyncerVWConfig), t.Name())
+	require.NoError(t, err)
+
+	startedSyncer := &StartedSyncerFixture{
+		SyncerConfig:          syncerConfig,
+		SyncerID:              syncerID,
+		SyncTargetClusterName: syncTargetClusterName,
+		DownstreamConfig:      downstreamConfig,
+		DownstreamKubeClient:  downstreamKubeClient,
+
+		SyncerVirtualWorkspaceConfig:   syncerVWConfig,
+		UpsyncerVirtualWorkspaceConfig: upsyncerVWConfig,
+	}
+
+	// The sync target becoming ready indicates the syncer is healthy and has
+	// successfully sent a heartbeat to kcp.
+	startedSyncer.WaitForClusterReady(ctx, t)
+
+	go wait.UntilWithContext(ctx, func(ctx context.Context) {
+		err = retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			syncTarget, err := kcpClusterClient.Cluster(syncerConfig.SyncTargetPath).WorkloadV1alpha1().SyncTargets().Get(ctx, syncerConfig.SyncTargetName, metav1.GetOptions{})
+			require.NoError(t, err)
+
+			authChecker := resourcesync.GVRAuthorizationChecker{
+				DownstreamKubeClient: downstreamKubeClient,
+				SyncTargetClient:     kcpSyncTargetClient.WorkloadV1alpha1().SyncTargets(),
+			}
+
+			var unauthorizedGVRs []string
+			for _, resource := range syncTarget.Status.SyncedResources {
+				for _, version := range resource.Versions {
+					gvr := schema.GroupVersionResource{
+						Group:    resource.Group,
+						Resource: resource.Resource,
+						Version:  version,
+					}
+					allowed, err := authChecker.CheckSSAR(ctx, gvr)
+					require.NoError(t, err)
+					if !allowed {
+						unauthorizedGVRs = append(unauthorizedGVRs, gvr.String())
+					}
+				}
+			}
+
+			return authChecker.PatchSyncTargetCondition(ctx, unauthorizedGVRs, syncTarget)
+		})
+		require.NoError(t, err)
+	}, 5*time.Second)
+
+	return startedSyncer
+}
```

Comment:
to be honest, after our conversation to create a fake syncer fixture I expected a 20 liner, maybe 50 lines, not 250. Why do we need such giant logic to keep a sync target alive? My gut feeling: this is yet another maintenance disaster.

Reply:
I removed the whole `GVRAuthorizationChecker` trick, and also factorized common code found in the current `Start` method. So I assume this is in a better shape now.
The number of lines specific to this fake Syncer start is effectively now back to about 50 lines.